maitake/src/wait/map.rs

```diff
@@ -458,25 +442,38 @@ impl<K: PartialEq, V> DataWaitQueue<K, V> {
     /// Wait to be woken up by this queue.
     ///
     /// This returns a [`Wait`] future that will complete when the task is
-    /// woken by a call to [`wake`] or [`wake_all`], or when the `DataWaitQueue` is
+    /// woken by a call to [`wake`] or [`wake_all`], or when the `WaitMap` is
     /// dropped.
     ///
     /// [`wake`]: Self::wake
     /// [`wake_all`]: Self::wake_all
     pub fn wait(&self, key: K) -> Wait<'_, K, V> {
-        Wait {
+        self.try_wait(key)
+            .expect("Keys should be unique in a `WaitMap`!")
+    }
+
+    pub fn try_wait(&self, key: K) -> Result<Wait<'_, K, V>, ()> {
+        // Check if key already exists
+        let mut mg = self.queue.lock();
+        let mut cursor = mg.cursor();
+        if cursor.find(|n| n.key == key).is_some() {
+            return Err(());
```

Comment:
do you think this should return the key value?

Reply:
Maybe! Probably couldn't hurt.